internal/operands/tekton-cleanup/reconcile.go

```diff
@@ -0,0 +1,195 @@
+		cleanupFuncs = append(cleanupFuncs, cleanupResource[pipeline.TaskList, pipeline.Task])         //nolint:staticcheck
+	}
+
+	var allResults []common.CleanupResult
+	for _, cleanupFunc := range cleanupFuncs {
+		cleanupResults, err := cleanupFunc(request)
+		if err != nil {
+			return nil, err
+		}
+		allResults = append(allResults, cleanupResults...)
+	}
+
+	return allResults, nil
+}
+
+func deprecateResource[L any, T any, PtrL interface {
+	*L
+	client.ObjectList
+}, PtrT interface {
+	*T
+	client.Object
+}](request *common.Request) error {
+	resources, err := common.ListOwnedResources[L, T, PtrL, PtrT](request, matchingLabelsOption(request.Instance))
+	if err != nil {
+		return fmt.Errorf("failed to list owned resources: %w", err)
+	}
+	for i := range resources {
+		resource := PtrT(&resources[i])
+		existingVal := resource.GetAnnotations()[tektonDeprecated]
+		if existingVal == "true" {
+			continue
+		}
+
+		resource.GetAnnotations()[tektonDeprecated] = "true"
+		if err := request.Client.Update(request.Context, resource); err != nil {
+			return fmt.Errorf("failed to update %s: %w", resource.GetObjectKind().GroupVersionKind().Kind, err)
+		}
+	}
+	return nil
+}
+
+func cleanupResource[L any, T any, PtrL interface {
+	*L
+	client.ObjectList
+}, PtrT interface {
+	*T
+	client.Object
+}](request *common.Request) ([]common.CleanupResult, error) {
+	resources, err := common.ListOwnedResources[L, T, PtrL, PtrT](request, matchingLabelsOption(request.Instance))
+	if err != nil {
+		return nil, fmt.Errorf("failed to list owned resources: %w", err)
+	}
+
+	results := make([]common.CleanupResult, 0, len(resources))
+	for i := range resources {
+		resource := PtrT(&resources[i])
+		cleanupResult, err := common.Cleanup(request, resource)
+		if err != nil {
+			return nil, fmt.Errorf("failed to cleanup resource: %w", err)
+		}
+		results = append(results, cleanupResult)
+	}
+	return results, nil
+}
+
+func matchingLabelsOption(ssp *ssp.SSP) client.MatchingLabelsSelector {
+	selector := labels.NewSelector().Add(
+		newLabelRequirementOrPanic(common.AppKubernetesManagedByLabel, selection.Equals, []string{common.AppKubernetesManagedByValue}),
+		newLabelRequirementOrPanic(common.AppKubernetesComponentLabel, selection.In, []string{common.AppComponentTektonPipelines.String(), common.AppComponentTektonTasks.String()}),
+		newLabelRequirementOrPanic(common.AppKubernetesNameLabel, selection.In, []string{operandPipelinesName, operandTasksName}),
+	)
+
+	if sspPartOf, exists := ssp.Labels[common.AppKubernetesPartOfLabel]; exists {
+		selector.Add(newLabelRequirementOrPanic(common.AppKubernetesPartOfLabel, selection.Equals, []string{sspPartOf}))
+	}
+
+	return client.MatchingLabelsSelector{Selector: selector}
+}
+
+func newLabelRequirementOrPanic(key string, op selection.Operator, vals []string) labels.Requirement {
+	requirement, err := labels.NewRequirement(key, op, vals)
+	if err != nil {
+		panic(fmt.Errorf("invalid requirement: %w", err))
```

Comment:
Shouldn't it return an `err` instead of panicking?

Reply:
I've added this helper function, because it is always called with arguments that are compile time constants. Any error returned by `labels.NewRequirement()` is a programmer's mistake. In my opinion it is ok to panic in this case.